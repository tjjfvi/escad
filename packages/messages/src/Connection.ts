
export interface Connection<T> {
  send: (value: T) => void,
  onMsg: (cb: (value: T) => void) => void,
  offMsg: (cb: (value: T) => void) => void,
}
