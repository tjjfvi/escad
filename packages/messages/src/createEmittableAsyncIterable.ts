
export const createEmittableAsyncIterable = <T>() => {
  let listeners = new Set<(value: T) => void>();

  return [
    (value: T) => {
      const oldListeners = listeners;
      listeners = new Set();
      oldListeners.forEach(fn => fn(value));
    },
    async function*(): AsyncIterable<T>{
      while(true)
        yield await new Promise(res => listeners.add(res))
    }
  ] as const;
}
