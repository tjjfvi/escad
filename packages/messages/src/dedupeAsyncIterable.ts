
export function dedupeAsyncIterable<T>(iterableFn: () => AsyncIterable<T>){
  let iterator: AsyncIterator<T> | undefined
  let curPromise: Promise<IteratorResult<T>> | undefined
  const dedupedIterable = {
    [Symbol.asyncIterator]: () => ({
      next: () => {
        iterator ??= iterableFn()[Symbol.asyncIterator]()
        curPromise ??= iterator.next().finally(() => curPromise = undefined)
        return curPromise
      },
    }),
  }
  const fn = async function*(){
    for await (const value of dedupedIterable)
      yield value
  }
  fn[Symbol.asyncIterator] = fn
  return fn
}
