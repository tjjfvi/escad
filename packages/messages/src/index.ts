
// @create-index {"mode":"*"}

export * from './Connection';
export * from './Messenger';
export * from './brandConnection';
export * from './childProcessConnection';
export * from './createMessenger';
export * from './filterConnection';
export * from './mapConnection';
export * from './noopConnection';
export * from './workerConnection';
export * from './wsConnection';

