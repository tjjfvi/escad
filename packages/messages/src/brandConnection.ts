
import { Connection } from "./Connection"
import { filterConnection } from "./filterConnection"
import { transformConnection } from "./transformConnection"

export const brandConnection =
  (connection: Connection<{ brand: string, value: unknown }, unknown>, brand: string): Connection<unknown> =>
    transformConnection(
      filterConnection(
        connection,
        (v): v is { brand: string, value: unknown } =>
          typeof v === "object" && !!v && "brand" in v && v["brand" as never] === brand && "value" in v,
      ),
      value => ({ brand, value }),
      ({ value }) => value,
    )
