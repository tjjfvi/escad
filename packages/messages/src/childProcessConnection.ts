
import type { ChildProcess } from "child_process"
import { Connection } from "./Connection"

/* istanbul ignore next: covered by types, difficult to test */
export const childProcessConnection = (childProcess: ChildProcess): Connection<unknown> => ({
  send: value => childProcess.send(value as never),
  onMsg: cb => {
    childProcess.on("message", cb)
    return () => childProcess.off("message", cb)
  },
  destroy: () => childProcess.kill(),
})
