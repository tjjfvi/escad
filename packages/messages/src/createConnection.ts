import { Connection } from "./Connection";

export const createConnection = <T>(send: (value: T) => void): Connection<T> & { emit: (value: T) => void } => {
  const listeners = new Set<(value: T) => void>();
  return {
    send,
    onMsg: cb => listeners.add(cb),
    offMsg: cb => listeners.delete(cb),
    emit: value => listeners.forEach(cb => cb(value)),
  }
}
