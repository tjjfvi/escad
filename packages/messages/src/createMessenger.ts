
import { Connection } from "./Connection";
import { Messenger, MessengerImpl, MessengerShape } from "./Messenger";

export const createMessenger = (
  <F extends MessengerShape, T extends MessengerShape>(
    impl: MessengerImpl<F, T>,
    connection: Connection<unknown>,
  ): Messenger<F, T> => {
    let idN = 0;
    let resolveMap = Object.create(null)
    let rejectMap = Object.create(null)
    let iteratorMap = Object.create(null)
    let destroyed = false;
    const other = new Proxy(Object.create(null), {
      get: (target, prop) => {
        if(destroyed)
          throw new Error("Attempted to make request on destroyed messenger")
        if(prop in target)
          return target[prop];
        if(typeof prop === "symbol")
          return;
        return target[prop] = (...args: any[]) => {
          const id = ++idN
          connection.send(["call", id, prop, ...args]);
          return Object.assign(
            recvPromise(id),
            {
              [Symbol.asyncIterator]: () => ({
                next: () => {
                  const id2 = ++idN
                  connection.send(["next", id, id2]);
                  return recvPromise(id2);
                },
              }),
            },
          );
        };
      },
    });
    (impl as any).req = other;
    (impl as any).destroy = () => {
      destroyed = true;
      resolveMap = Object.create(null);
      rejectMap = Object.create(null);
      iteratorMap = Object.create(null);
      connection.offMsg(handler);
      connection.destroy?.();
    };
    const handler = (msg: unknown) => {
      if(!msg || !(msg instanceof Array))
        return;
      const [kind, id, key, ...args] = msg;
      if(typeof id !== "number" || typeof key !== "number" && typeof key !== "string" && key !== null)
        return;
      switch(kind) {
        case "call": {
          const result = impl[key]?.(...args) ?? Promise.reject(new Error(`No method with the key "${key}"`));
          if("then" in result)
            sendPromise(id, result);
          else if(Symbol.asyncIterator in result)
            iteratorMap[id] = result[Symbol.asyncIterator]();
          return;
        }
        case "next": {
          sendPromise(+key, iteratorMap[id].next());
          return;
        }
        case "resolve": {
          if(id in resolveMap)
            resolveMap[id](args[0]);
          delete resolveMap[id];
          return;
        }
        case "reject": {
          if(id in rejectMap)
            rejectMap[id](args[0]);
          delete rejectMap[id];
          return;
        }
      }
    }
    connection.onMsg(handler);
    return impl as any;

    function recvPromise(id: number){
      return new Promise((resolve, reject) => {
        resolveMap[id] = resolve;
        rejectMap[id] = reject
      });
    }

    function sendPromise(id: number, promise: Promise<unknown>){
      promise.then(
        value =>
          !destroyed && connection.send(["resolve", id, null, value]),
        value =>
          !destroyed && connection.send(["reject", id, null, value]),
      )
    }
  }
)
