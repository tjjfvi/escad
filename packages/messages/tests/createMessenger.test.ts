import { createMessenger, MessengerImpl, noopConnection } from "../src"

type TestMessenger = {
  promiseResolve: <T>(value: T) => Promise<T>,
  promiseReject: <T>(value: T) => Promise<void>,
  reallyDumbPromiseResolve: <T>(value: T, count: number) => Promise<T>,
  asyncIterable: (values: number[]) => AsyncIterable<number>,
}

const testMessengerImpl: MessengerImpl<TestMessenger, TestMessenger> = {
  promiseResolve: value => Promise.resolve(value),
  promiseReject: value => Promise.reject(value),
  async *asyncIterable(values){
    for(const value of values) {
      if(value === -1)
        throw new Error("The value was `-1`")
      yield value
    }
  },
  async reallyDumbPromiseResolve(value, count){
    if(count <= 0)
      return value
    return this.req.reallyDumbPromiseResolve(value, count - 1)
  },
}

test("", async () => {
  const [a, b] = noopConnection<unknown>()
  const fn = jest.fn()
  a.onMsg(x => fn("b->a", x))
  b.onMsg(x => fn("a->b", x))
  const msgrA = createMessenger(testMessengerImpl, a)
  const msgrB = createMessenger(testMessengerImpl, b)
  expect(await msgrA.promiseResolve(-1)).toBe(-1)
  expect(msgrA.promiseReject(-1)).rejects.toBe(-1)
  expect(await msgrA.req.promiseResolve(0)).toBe(0)
  expect(msgrA.req.promiseReject(1)).rejects.toBe(1)
  expect(await msgrB.req.promiseResolve(2)).toBe(2)
  for await (const value of msgrA.req.asyncIterable([3, 4, 5]))
    fn(value)
  await expect((async () => {
    for await (const value of msgrA.req.asyncIterable([6, -1, -2]))
      fn(value)
  })()).rejects.toMatchSnapshot()
  expect(await msgrA.reallyDumbPromiseResolve(6, 10)).toBe(6)
  b.send(1234)
  b.send(null)
  b.send({})
  b.send([])
  b.send(["missingno", 1, 2, 3])
  b.send(["call", null, 1, 1])
  b.send(["call", 1, {}, 1])
  b.send(["call", 404, "missingno"])
  expect(fn.mock.calls).toMatchSnapshot()
})
