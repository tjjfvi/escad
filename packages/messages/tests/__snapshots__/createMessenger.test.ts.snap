// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    "a->b",
    Array [
      "call",
      1,
      "promiseResolve",
      -1,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      1,
      null,
      -1,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      2,
      "promiseResolve",
      0,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      2,
      null,
      0,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      1,
      "promiseResolve",
      1,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      1,
      null,
      1,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      3,
      "bouncingPromiseResolve",
      2,
      10,
    ],
  ],
  Array [
    "b->a",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      10,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      2,
      "bouncingPromiseResolve",
      2,
      9,
    ],
  ],
  Array [
    "a->b",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      9,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      4,
      "bouncingPromiseResolve",
      2,
      8,
    ],
  ],
  Array [
    "b->a",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      8,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      3,
      "bouncingPromiseResolve",
      2,
      7,
    ],
  ],
  Array [
    "a->b",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      7,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      5,
      "bouncingPromiseResolve",
      2,
      6,
    ],
  ],
  Array [
    "b->a",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      6,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      4,
      "bouncingPromiseResolve",
      2,
      5,
    ],
  ],
  Array [
    "a->b",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      5,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      6,
      "bouncingPromiseResolve",
      2,
      4,
    ],
  ],
  Array [
    "b->a",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      4,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      5,
      "bouncingPromiseResolve",
      2,
      3,
    ],
  ],
  Array [
    "a->b",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      3,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      7,
      "bouncingPromiseResolve",
      2,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      6,
      "bouncingPromiseResolve",
      2,
      1,
    ],
  ],
  Array [
    "a->b",
    Array [
      "event",
      -1,
      "promiseResolveBounced",
      2,
      1,
    ],
  ],
  Array [
    "a->b",
    Array [
      "call",
      8,
      "bouncingPromiseResolve",
      2,
      0,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      8,
      null,
      2,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      6,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      7,
      null,
      2,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      5,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      6,
      null,
      2,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      4,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      5,
      null,
      2,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      3,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      4,
      null,
      2,
    ],
  ],
  Array [
    "a->b",
    Array [
      "resolve",
      2,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    Array [
      "resolve",
      3,
      null,
      2,
    ],
  ],
  Array [
    "b->a",
    1234,
  ],
  Array [
    "b->a",
    null,
  ],
  Array [
    "b->a",
    Object {},
  ],
  Array [
    "b->a",
    Array [],
  ],
  Array [
    "b->a",
    Array [
      "missingno",
      1,
      2,
      3,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      null,
      1,
      1,
    ],
  ],
  Array [
    "b->a",
    Array [
      "call",
      1,
      Object {},
      1,
    ],
  ],
]
`;
